src/System.Web.OData/OData/Formatter/EdmLibHelpers.cs

```diff
@@ -787,8 +785,16 @@ public static IEnumerable<IEdmStructuralProperty> GetConcurrencyProperties(this
             Contract.Assert(model != null);
             Contract.Assert(navigationSource != null);
 
+            // Ensure that concurrency properties cache is attached to model as an annotation to avoid expensive calculations each time
+            ConcurrentDictionary<IEdmNavigationSource, IEnumerable<IEdmStructuralProperty>> concurrencyProperties = model.GetAnnotationValue<ConcurrentDictionary<IEdmNavigationSource, IEnumerable<IEdmStructuralProperty>>>(model);
+            if (concurrencyProperties == null)
+            {
+                concurrencyProperties = new ConcurrentDictionary<IEdmNavigationSource, IEnumerable<IEdmStructuralProperty>>();
+                model.SetAnnotationValue(model, concurrencyProperties);
+            }
+
             IEnumerable<IEdmStructuralProperty> cachedProperties;
-            if (_concurrencyProperties != null && _concurrencyProperties.TryGetValue(navigationSource, out cachedProperties))
+            if (concurrencyProperties != null && concurrencyProperties.TryGetValue(navigationSource, out cachedProperties))
```

Comment:
>concurrencyProperties [](start = 16, length = 21)

This will no longer ever be null at this point, right? If it was null in line 790 you would have created it in 792.